File: src/provider/FreeVersionProvider.java
package provider;

import com.intellij.psi.xml.XmlTag;
import domain.Dependency;

import java.util.Arrays;

/**
 *
 */
public class FreeVersionProvider {
    private static final String VERSION_SUFFIX = ".version";
    private static final char DASH = '-';
    private static final int MAX_ATTEMPTS = 20;

    private final PropertiesProvider propertiesProvider = new PropertiesProvider();

    /**
     *
     * @param project
     * @param dependency
     * @return
     */
    public String getFreeVersion(final XmlTag project, final Dependency dependency) {
        final var properties = propertiesProvider.provide(project);
        String candidate = dependency.getArtifactId() + VERSION_SUFFIX;
        if (properties != null) {
            if (isPropertyInUse(properties, candidate)) {
                candidate = fallbackWithFullName(properties, dependency);
            }
        }
        return candidate;
    }

    private String fallbackWithFullName(final XmlTag properties, final Dependency dependency) {
        String candidate = dependency.getGroupId() + DASH + dependency.getArtifactId() + VERSION_SUFFIX;
        int attempt = 1;
        while (isPropertyInUse(properties, candidate) && attempt < MAX_ATTEMPTS) {
            candidate = dependency.getGroupId() + DASH + dependency.getArtifactId() + DASH + (attempt+1) + VERSION_SUFFIX;
            attempt++;
        }
        if (isPropertyInUse(properties, candidate)) {
            candidate = null;
        }
        return candidate;
    }

    private boolean isPropertyInUse(final XmlTag properties, final String property) {
        final var name = property.toLowerCase();
        return Arrays.stream(properties.getChildren())
                .filter(XmlTag.class::isInstance)
                .map(XmlTag.class::cast)
                .anyMatch(xmlTag -> name.equals(xmlTag.getName().toLowerCase()));
    }
}

File: src/provider/FreeVersionProvider.kt
package provider

import com.intellij.psi.xml.XmlTag
import domain.Dependency

class FreeVersionProvider {
    companion object {
        private const val VERSION_SUFFIX = ".version"
        private const val DASH = '-'
        private const val MAX_ATTEMPTS = 20
    }

    private val propertiesProvider = PropertiesProvider()

    fun getFreeVersion(project: XmlTag, dependency: Dependency): String? {
        val properties = propertiesProvider.provide(project)
        val candidate = dependency.artifactId + VERSION_SUFFIX
        return if (properties == null || !isPropertyInUse(properties, candidate)) {
            candidate
        } else {
            fallbackWithFullName(properties, dependency)
        }
    }

    private fun fallbackWithFullName(properties: XmlTag, dependency: Dependency): String? {
        val candidate = dependency.groupId + DASH + dependency.artifactId + VERSION_SUFFIX
        return if (!isPropertyInUse(properties, candidate)) {
            candidate
        } else {
            val prefix = dependency.groupId + DASH + dependency.artifactId + DASH
            fallbackWithFullNameAndAttempt(properties, prefix, 1)
        }
    }

    private fun fallbackWithFullNameAndAttempt(properties: XmlTag, prefix: String, attempt: Int): String? =
            if (attempt <= MAX_ATTEMPTS) {
                val candidate = prefix + (attempt+1) + VERSION_SUFFIX
                if (!isPropertyInUse(properties, candidate)) {
                    candidate
                } else {
                    fallbackWithFullNameAndAttempt(properties, prefix, attempt+1)
                }
            } else {
                null
            }

    private fun isPropertyInUse(properties: XmlTag, property: String) =
            properties.children
                    .filter { XmlTag::class.java.isInstance(it) }
                    .map { XmlTag::class.java.cast(it) }
                    .any { property.toLowerCase() == it.name.toLowerCase() }
}
